### ISSUE TITLE
arping command count error

### ISSUE THREAD

**Issue by firexinghe** [iputils/iputils#392]
Return value error, receive and send values do not match 

20211215 version

# arping 10.82.188.10 -I eth0 -s 10.82.188.11 -w 3
ARPING 10.82.188.10 from 10.82.188.11 eth0
Unicast reply from 10.82.188.10 [xx:xx:xx:xx]  1.176ms
Unicast reply from 10.82.188.10 [xx:xx:xx:xx]  0.686ms
Unicast reply from 10.82.188.10 [xx:xx:xx:xx]  0.698ms
Sent 4 probes (1 broadcast(s))
Received 3 response(s)

# echo $?
1

**Comment by eaglegai** [comment]
maybe commit e594ca52 cause this

**Comment by pevik** [comment]
It looks like before e594ca52afde89746b7d79c875fe9d6aea1850ac `-w deadline` just didn't work. But indeed with proper implementation it always miss the latest reply. I need to have closer look @kerolasa any idea why?

**Comment by nmeyerhans** [comment]
The exit code definitely looks wrong (this arping appears to have been successful, so should have exited with status 0)

However, the packet count doesn't necessarily look wrong to me. `-w Z` without `-c`, as documented in the manpage, instructs arping to exit after Z responses are processed, regardless of how many packets have been sent. It's entirely possible that a 4th packet was in flight when the timer triggered, and that arping never processed a reply.

**Comment by pevik** [comment]
Agree. Should we `-w Z` without `-c` just ignore if the latest one is missing? Or ignore missing replies if even more than one is missing?

**Comment by nmeyerhans** [comment]
I think arping should behave as we document ping to behave in this case:
>       If ping does not receive any reply packets at all it will exit with code 1. If a packet count and deadline are
>       both specified, and fewer than count packets are received by the time the deadline has arrived, it will also
>       exit with code 1. On other error it exits with code 2. Otherwise it exits with code 0. This makes it possible
>       to use the exit code to see if a host is alive or not.

So in this case, with no packet count provided, arping should exit 0.

However, ping is a little bit funny, too, and I think we need to dig in a little deeper to the deadline handling and packet accounting code in both programs.

With arping, we see the following:
```
$ time sudo ./arping 192.168.1.1 -I wlp0s20f3 -s 192.168.1.214 -w3
ARPING 192.168.1.1 from 192.168.1.214 wlp0s20f3
Unicast reply from 192.168.1.1 [C0:56:27:C7:CD:76]  4.419ms
Unicast reply from 192.168.1.1 [C0:56:27:C7:CD:76]  4.621ms
Unicast reply from 192.168.1.1 [C0:56:27:C7:CD:76]  4.203ms
Sent 4 probes (1 broadcast(s))
Received 3 response(s)
sudo ./arping 192.168.1.1 -I wlp0s20f3 -s 192.168.1.214 -w3  0.01s user 0.02s system 1% cpu 3.045 total
```
We can confirm with tcpdump that 4 packets are sent. There are 4 responses as well, but the program apparently exits before processing the final reply. Note that the complete runtime was reported as 3.045s.

Compared with ping, where we see:
```
$ time sudo ./ping/ping 192.168.1.1 -I 192.168.1.214 -w3         
PING 192.168.1.1 (192.168.1.1) from 192.168.1.214 : 56(84) bytes of data.
64 bytes from 192.168.1.1: icmp_seq=1 ttl=64 time=4.38 ms
64 bytes from 192.168.1.1: icmp_seq=2 ttl=64 time=3.27 ms
64 bytes from 192.168.1.1: icmp_seq=3 ttl=64 time=3.82 ms

--- 192.168.1.1 ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 2005ms
rtt min/avg/max/mdev = 3.267/3.823/4.383/0.455 ms
sudo ./ping/ping 192.168.1.1 -I 192.168.1.214 -w3  0.01s user 0.01s system 0% cpu 3.022 total
```

We can confirm with tcpdump that 3 packets are sent, and replies for each are processed. ping claims to have taken 2005ms, yet the shell measured 3.022s in wallclock time elapsing. So neither seems to be handling the deadline option entirely correctly.

**Comment by eaglegai** [comment]
from tcpdump we know, send 4 packages, and receive 4 reply packages  in fact，but arping always drop the final，how about we add checking like this：
                        case POLLFD_TIMEOUT:
+                               if (recv_pack
+                                   (ctl, packet, s, (struct sockaddr_ll *)&from) != FINAL_PACKS && ctl->received != 0)
+                                       ctl->received++;
                                exit_loop = 1;
                                break;

**Comment by nmeyerhans** [comment]
I opened #395 to implement what I believe should be the correct behavior in this case.

In general, I don't think it makes sense to wait for in-flight packets if we've asked arping to exit after a certain amount of time. The docs indicate that `-w` means "arping exits regardless of how many packets have been sent or received" and I think that's what it should do. Note that ping also will leave some packets in flight if the link latency is high enough.

If the details of arping's interval and the given deadline leave a packet in flight when the deadline expires, that's fine. We should report on the replies we've seen, and exit with a meaningful status.

When thinking about this, it helped me to add some instrumentation to arping so I could see exactly what events were firing and when.  You can see some of that output below.  What this is showing is that the 4th packet is being sent at the same time as the deadline timer expires, which is what leaves the 4th packet in flight when we exit. Logically, I think this behavior is correct. With a deadline of 3s and the default 1s transmission interval, we expect to send a packet at t=0, t=1, t=2, and t=3. It happens that we also expect to exit at t=3, which means we don't stick around long enough for the reply to come in. You could argue that we should take steps to ensure that the deadline timer is always processed before the interval timer, but aside from munging the deadline or interval values themselves, I don't think we can do this. Even if we could, I'm not sure how reliable it would be, and I don't think it's a good idea anyway, since the values provided by the user should be respected.

```text
$ sudo ./arping 192.168.1.2 -I wlp0s20f3 -w3
ARPING 192.168.1.2 from 192.168.1.215 wlp0s20f3
Sending first packet
poll: none
poll: interval timer event: sending a packet
poll: none
poll: none
poll: none
poll: interval timer event: sending a packet
poll: none
poll: none
poll: none
poll: interval timer event: sending a packet
poll: timeout timer event
poll: none
Sent 4 probes (4 broadcast(s))
Received 0 response(s)
Returning 1
$ sudo ./arping 192.168.1.1 -I wlp0s20f3 -w3
ARPING 192.168.1.1 from 192.168.1.215 wlp0s20f3
Sending first packet
poll: none
poll: none
poll: none
poll: socket event
Wed Feb 16 22:57:37 2022
: Unicast reply from 192.168.1.1 [C0:56:27:C7:CD:76]  3.430ms
poll: none
poll: interval timer event: sending a packet
poll: none
poll: none
poll: none
poll: none
poll: none
poll: socket event
Wed Feb 16 22:57:38 2022
: Unicast reply from 192.168.1.1 [C0:56:27:C7:CD:76]  4.052ms
poll: none
poll: interval timer event: sending a packet
poll: none
poll: none
poll: none
poll: none
poll: none
poll: socket event
Wed Feb 16 22:57:39 2022
: Unicast reply from 192.168.1.1 [C0:56:27:C7:CD:76]  4.087ms
poll: none
poll: interval timer event: sending a packet
poll: timeout timer event
poll: none
Sent 4 probes (1 broadcast(s))
Received 3 response(s)
Returning 0
```
